firefox-ios/Client/Frontend/TrackingProtection/CertificatesHandler.swift

```diff
@@ -70,18 +70,22 @@ class CertificateDelegate: NSObject, URLSessionDelegate {
         self.completion = completion
     }
 
-    func urlSession(_ session: URLSession, didReceive
-                    challenge: URLAuthenticationChallenge,
-                    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
-        if let serverTrust = challenge.protectionSpace.serverTrust {
-            let certificatesHandler = CertificatesHandler(serverTrust: serverTrust)
-            self.completion(certificatesHandler.handleCertificates())
-            // Use the server trust
-            completionHandler(.useCredential, URLCredential(trust: serverTrust))
-        } else {
-            // Call the completion handler with nil if serverTrust is not available
-            self.completion(nil)
-            completionHandler(.cancelAuthenticationChallenge, nil)
+    func urlSession(
+        _ session: URLSession,
+        didReceive challenge: URLAuthenticationChallenge
+    ) async -> (
+        URLSession.AuthChallengeDisposition,
+        URLCredential?
+    ) {
+        return await withCheckedContinuation { continuation in
```

Comment:
Do we need here `withCheckedContinuation` ? the code seems synchronous.

Reply:
The `URLSession` delegate method `urlSession(_:didReceive:completionHandler:)` is inherently asynchronous. It is designed to handle authentication challenges that occur during a URL session task. The method expects you to perform some asynchronous work (like validating server trust) and then call the provided completion handler when done. 

What's changed in Swift 6 is that the `completionHandler` needs to be now `@Sendable`. This is the new function signature: `func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping @Sendable (URLSession.AuthChallengeDisposition, URLCredential?) -> Void)`.

To make things simpler I chose the concurrency approach by conforming to: `func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge) async -> (URLSession.AuthChallengeDisposition, URLCredential?)` that uses the async/await concurrency model and used `withCheckedContinuation` as a bridge between the modern async/await concurrency model and traditional completion handler-based asynchronous APIs.

Reply:
True i think that approach is great. But i'm not understanding the `withCheckedContinuation` because i see in the closure of the continuation there is no method that pass a completion handler. Sorry to ask again, but i may miss something under the hood, like some Thread implication.

I can compile removing the continuation.
```swift
func urlSession(
        _ session: URLSession,
        didReceive challenge: URLAuthenticationChallenge
    ) async -> (
        URLSession.AuthChallengeDisposition,
        URLCredential?
    ) {
        if let serverTrust = challenge.protectionSpace.serverTrust {
            let certificatesHandler = CertificatesHandler(serverTrust: serverTrust)
            self.completion(certificatesHandler.handleCertificates())
            return (.useCredential, URLCredential(trust: serverTrust))
        } else {
            self.completion(nil)
            return (.cancelAuthenticationChallenge, nil)
        }
    }
```

Reply:
Yes, at a closer look, we don't have any asynchronous operations and indeed we don't need `withCheckedContinuation`. Thanks for the detailed approach, I'll update the code.

Reply:
Thanks @PARAIPAN9 🙏